### PR TITLE
Update Travis, Jenkins and local environment setup to use the same me…

### PR DESCRIPTION
**Pull request by lwgordonimos**
…chanism to install dependencies.

@mhidas this builds on your (good) idea to base the build dependencies on stages, and moves the actual installation into a script which can be *shared* between anything which wants to setup an environment to run aodndata.

The problem this solves is that the local setup_virtualenv.sh, the travis.yml and now the Jenkinsfile are at a high risk of diverging and unnecessarily replicating what is exactly the same thing: automatically installing AODN dependencies in addition to PyPI dependencies.

### CODE REVIEW

File: scripts/install_dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+set -ex
+
+STAGE=${STAGE:-prod}
```
Comment (mhidas) [inline scripts/install_dependencies.sh]: Just for convenience, STAGE could also be an optional command-line arg?

**Comment by mhidas** [comment]
Hmm... I just tried  the `setup_virtualenv.sh` script, and I got this...
```
##### Creating/updating virtual environment #####
Running virtualenv with interpreter /usr/bin/python2
New python executable in /sw/chef/src/python-aodndata/python-aodndata-virtualenv/bin/python2
Also creating executable in /sw/chef/src/python-aodndata/python-aodndata-virtualenv/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
+ STAGE=prod
+ AODNFETCHER_URL=git+https://github.com/aodn/python-aodnfetcher.git@master
+ AODNCORE_URL='jenkins://imos-binary/aodncore_prod?pattern=^.*\.whl$'
+ NCWRITER_URL='jenkins://imos-binary/ncwriter_prod?pattern=^.*\.whl$'
+ CC_PLUGIN_IMOS_URL='jenkins://imos-binary/cc_plugin_imos_prod?pattern=^.*\.whl$'
+ WHEEL_CACHE_DIR=.python-aodndata-download-cache
+ echo '##### Installing dependencies into virtual environment #####'
##### Installing dependencies into virtual environment #####
+ pip install git+https://github.com/aodn/python-aodnfetcher.git@master
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
Collecting git+https://github.com/aodn/python-aodnfetcher.git@master
  Cloning https://github.com/aodn/python-aodnfetcher.git (to revision master) to /tmp/pip-req-build-89nZPh
  Running command git clone -q https://github.com/aodn/python-aodnfetcher.git /tmp/pip-req-build-89nZPh
Collecting boto3>=1.9.156 (from aodnfetcher==0.4)
  Downloading https://files.pythonhosted.org/packages/3b/a2/0d88f3437916f4ec03e2964d614207d710472e2a1eedcec732392c4fede3/boto3-1.9.160-py2.py3-none-any.whl (128kB)
     |████████████████████████████████| 133kB 3.1MB/s 
Collecting fasteners==0.14.1 (from aodnfetcher==0.4)
  Downloading https://files.pythonhosted.org/packages/14/3a/096c7ad18e102d4f219f5dd15951f9728ca5092a3385d2e8f79a7c1e1017/fasteners-0.14.1-py2.py3-none-any.whl
Collecting requests<2.20.0,>=2.12.1 (from aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/65/47/7e02164a2a3db50ed6d8a6ab1d6d60b69c4c3fdf57a284257925dfc12bda/requests-2.19.1-py2.py3-none-any.whl
Collecting botocore<1.13.0,>=1.12.160 (from boto3>=1.9.156->aodnfetcher==0.4)
  Downloading https://files.pythonhosted.org/packages/67/b8/f55d7c466681f5b6a75314809d0899a4a4d4697131dfc9b52f76c492c8f7/botocore-1.12.160-py2.py3-none-any.whl (5.5MB)
     |████████████████████████████████| 5.5MB 72.3MB/s 
Collecting s3transfer<0.3.0,>=0.2.0 (from boto3>=1.9.156->aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/d7/de/5737f602e22073ecbded7a0c590707085e154e32b68d86545dcc31004c02/s3transfer-0.2.0-py2.py3-none-any.whl
Collecting jmespath<1.0.0,>=0.7.1 (from boto3>=1.9.156->aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/83/94/7179c3832a6d45b266ddb2aac329e101367fbdb11f425f13771d27f225bb/jmespath-0.9.4-py2.py3-none-any.whl
Collecting six (from fasteners==0.14.1->aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
Collecting monotonic>=0.1 (from fasteners==0.14.1->aodnfetcher==0.4)
  Downloading https://files.pythonhosted.org/packages/ac/aa/063eca6a416f397bd99552c534c6d11d57f58f2e94c14780f3bbf818c4cf/monotonic-1.5-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests<2.20.0,>=2.12.1->aodnfetcher==0.4)
  Downloading https://files.pythonhosted.org/packages/60/75/f692a584e85b7eaba0e03827b3d51f45f571c2e793dd731e598828d380aa/certifi-2019.3.9-py2.py3-none-any.whl (158kB)
     |████████████████████████████████| 163kB 58.1MB/s 
Collecting chardet<3.1.0,>=3.0.2 (from requests<2.20.0,>=2.12.1->aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl
Collecting urllib3<1.24,>=1.21.1 (from requests<2.20.0,>=2.12.1->aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/bd/c9/6fdd990019071a4a32a5e7cb78a1d92c53851ef4f56f62a3486e6a7d8ffb/urllib3-1.23-py2.py3-none-any.whl
Collecting idna<2.8,>=2.5 (from requests<2.20.0,>=2.12.1->aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/4b/2a/0276479a4b3caeb8a8c1af2f8e4355746a97fab05a372e4a2c6a6b876165/idna-2.7-py2.py3-none-any.whl
Collecting python-dateutil<3.0.0,>=2.1; python_version >= "2.7" (from botocore<1.13.0,>=1.12.160->boto3>=1.9.156->aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/41/17/c62faccbfbd163c7f57f3844689e3a78bae1f403648a6afb1d0866d87fbb/python_dateutil-2.8.0-py2.py3-none-any.whl
Collecting docutils>=0.10 (from botocore<1.13.0,>=1.12.160->boto3>=1.9.156->aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/50/09/c53398e0005b11f7ffb27b7aa720c617aba53be4fb4f4f3f06b9b5c60f28/docutils-0.14-py2-none-any.whl
Collecting futures<4.0.0,>=2.2.0; python_version == "2.6" or python_version == "2.7" (from s3transfer<0.3.0,>=0.2.0->boto3>=1.9.156->aodnfetcher==0.4)
  Using cached https://files.pythonhosted.org/packages/2d/99/b2c4e9d5a30f6471e410a146232b4118e697fa3ffc06d6a65efde84debd0/futures-3.2.0-py2-none-any.whl
Building wheels for collected packages: aodnfetcher
  Building wheel for aodnfetcher (setup.py) ... done
  Stored in directory: /tmp/pip-ephem-wheel-cache-ur99aB/wheels/13/16/e6/76d03b2af4e2914a2af542df579fb3fd593312ee10606f1ccb
Successfully built aodnfetcher
Installing collected packages: urllib3, jmespath, six, python-dateutil, docutils, botocore, futures, s3transfer, boto3, monotonic, fasteners, certifi, chardet, idna, requests, aodnfetcher
Successfully installed aodnfetcher-0.4 boto3-1.9.160 botocore-1.12.160 certifi-2019.3.9 chardet-3.0.4 docutils-0.14 fasteners-0.14.1 futures-3.2.0 idna-2.7 jmespath-0.9.4 monotonic-1.5 python-dateutil-2.8.0 requests-2.19.1 s3transfer-0.2.0 six-1.12.0 urllib3-1.23
++ python -c 'import json, sys; print json.load(sys.stdin)['\''jenkins://imos-binary/cc_plugin_imos_prod?pattern=^.*\.whl$'\'']['\''local_file'\'']'
++ aodnfetcher -c .python-aodndata-download-cache 'jenkins://imos-binary/cc_plugin_imos_prod?pattern=^.*\.whl$'
+ pip install cc_plugin_imos-1.3.6-py2.py3-none-any.whl
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
Processing ./cc_plugin_imos-1.3.6-py2.py3-none-any.whl
Collecting netCDF4>=1.2.4 (from cc-plugin-imos==1.3.6)
  Using cached https://files.pythonhosted.org/packages/ed/20/780fb3c87db715a64b7b8f506fffc7dee169f38a56a158add2db695b2ad2/netCDF4-1.5.1.2-cp27-cp27mu-manylinux1_x86_64.whl
Collecting compliance-checker>=2.0.0 (from cc-plugin-imos==1.3.6)
Collecting numpy>=1.7 (from netCDF4>=1.2.4->cc-plugin-imos==1.3.6)
  Downloading https://files.pythonhosted.org/packages/1f/c7/198496417c9c2f6226616cff7dedf2115a4f4d0276613bab842ec8ac1e23/numpy-1.16.4-cp27-cp27mu-manylinux1_x86_64.whl (17.0MB)
     |████████████████████████████████| 17.0MB 3.0MB/s 
Collecting cftime (from netCDF4>=1.2.4->cc-plugin-imos==1.3.6)
  Using cached https://files.pythonhosted.org/packages/c8/9f/dc02843e350550bc2a0522d36314e995891a8715934359dbe842dcc1920d/cftime-1.0.3.4-cp27-cp27mu-manylinux1_x86_64.whl
Collecting isodate>=0.5.4 (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Using cached https://files.pythonhosted.org/packages/9b/9f/b36f7774ff5ea8e428fdcfc4bb332c39ee5b9362ddd3d40d9516a55221b2/isodate-0.6.0-py2.py3-none-any.whl
Requirement already satisfied: setuptools>=15.0 in ./python-aodndata-virtualenv/lib/python2.7/site-packages (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6) (41.0.1)
Collecting Jinja2>=2.7.3 (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Using cached https://files.pythonhosted.org/packages/1d/e7/fd8b501e7a6dfe492a433deb7b9d833d39ca74916fa8bc63dd1a4947a671/Jinja2-2.10.1-py2.py3-none-any.whl
Collecting functools32==3.2.3-2; python_version < "3.2" (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
Collecting cf-units>=2 (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
Collecting lxml>=3.2.1 (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Using cached https://files.pythonhosted.org/packages/89/51/a8a6cdb8a084d32dbc9bda94623dc35310ae2002be57de8702a1703c0026/lxml-4.3.3-cp27-cp27mu-manylinux1_x86_64.whl
Collecting OWSLib>=0.8.3 (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
Collecting regex>=2017.07.28 (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Downloading https://files.pythonhosted.org/packages/29/a4/858a23347cd5e5f238524b5a0a9d52e96aee2ed7ab37534134314a767c87/regex-2019.06.02.tar.gz (651kB)
     |████████████████████████████████| 655kB 56.3MB/s 
Collecting pendulum>=1.2.4 (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
Collecting pygeoif>=0.6 (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
Requirement already satisfied: requests>=2.2.1 in ./python-aodndata-virtualenv/lib/python2.7/site-packages (from compliance-checker>=2.0.0->cc-plugin-imos==1.3.6) (2.19.1)
Requirement already satisfied: six in ./python-aodndata-virtualenv/lib/python2.7/site-packages (from isodate>=0.5.4->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6) (1.12.0)
Collecting MarkupSafe>=0.23 (from Jinja2>=2.7.3->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Using cached https://files.pythonhosted.org/packages/fb/40/f3adb7cf24a8012813c5edb20329eb22d5d8e2a0ecf73d21d6b85865da11/MarkupSafe-1.1.1-cp27-cp27mu-manylinux1_x86_64.whl
Collecting cython (from cf-units>=2->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Downloading https://files.pythonhosted.org/packages/38/d4/b8aaae645d1d33dfca8f95018fbcafdf5a882b80693a34165ff0b5c26aba/Cython-0.29.10-cp27-cp27mu-manylinux1_x86_64.whl (2.0MB)
     |████████████████████████████████| 2.0MB 62.7MB/s 
Collecting pyproj (from OWSLib>=0.8.3->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Downloading https://files.pythonhosted.org/packages/72/e4/a523c53e6b929bb4e488fdae13f2106e7ccc4e273f9cb473c16f406b534d/pyproj-2.2.0-cp27-cp27mu-manylinux1_x86_64.whl (11.2MB)
     |████████████████████████████████| 11.2MB 65.9MB/s 
Collecting pytz (from OWSLib>=0.8.3->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Downloading https://files.pythonhosted.org/packages/3d/73/fe30c2daaaa0713420d0382b16fbb761409f532c56bdcc514bf7b6262bb6/pytz-2019.1-py2.py3-none-any.whl (510kB)
     |████████████████████████████████| 512kB 59.3MB/s 
Requirement already satisfied: python-dateutil>=1.5 in ./python-aodndata-virtualenv/lib/python2.7/site-packages (from OWSLib>=0.8.3->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6) (2.8.0)
Collecting typing<4.0,>=3.6; python_version < "3.5" (from pendulum>=1.2.4->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Using cached https://files.pythonhosted.org/packages/cc/3e/29f92b7aeda5b078c86d14f550bf85cff809042e3429ace7af6193c3bc9f/typing-3.6.6-py2-none-any.whl
Collecting pytzdata>=2018.3 (from pendulum>=1.2.4->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Using cached https://files.pythonhosted.org/packages/a7/90/c9a148b4c98ccf88765ab7399debd38a33660109c9ba23706f560109b6e1/pytzdata-2019.1-py2.py3-none-any.whl
Requirement already satisfied: certifi>=2017.4.17 in ./python-aodndata-virtualenv/lib/python2.7/site-packages (from requests>=2.2.1->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6) (2019.3.9)
Requirement already satisfied: chardet<3.1.0,>=3.0.2 in ./python-aodndata-virtualenv/lib/python2.7/site-packages (from requests>=2.2.1->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6) (3.0.4)
Requirement already satisfied: urllib3<1.24,>=1.21.1 in ./python-aodndata-virtualenv/lib/python2.7/site-packages (from requests>=2.2.1->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6) (1.23)
Requirement already satisfied: idna<2.8,>=2.5 in ./python-aodndata-virtualenv/lib/python2.7/site-packages (from requests>=2.2.1->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6) (2.7)
Collecting aenum; python_version < "3.6" (from pyproj->OWSLib>=0.8.3->compliance-checker>=2.0.0->cc-plugin-imos==1.3.6)
  Downloading https://files.pythonhosted.org/packages/44/c4/7cda7e7e71e18ad999cb9d23f9969f818af1ed6be71d18db7963680b8320/aenum-2.1.2-py2-none-any.whl
Building wheels for collected packages: regex
  Building wheel for regex (setup.py) ... done
  Stored in directory: /home/mhidas/.cache/pip/wheels/d6/a1/71/76ede1c1ae34dea05829f8cb52f31e513b887d2319e67da62a
Successfully built regex
Installing collected packages: numpy, cftime, netCDF4, isodate, MarkupSafe, Jinja2, functools32, cython, cf-units, lxml, aenum, pyproj, pytz, OWSLib, regex, typing, pytzdata, pendulum, pygeoif, compliance-checker, cc-plugin-imos
Successfully installed Jinja2-2.10.1 MarkupSafe-1.1.1 OWSLib-0.17.1 aenum-2.1.2 cc-plugin-imos-1.3.6 cf-units-2.1.1 cftime-1.0.3.4 compliance-checker-4.2.2 cython-0.29.10 functools32-3.2.3.post2 isodate-0.6.0 lxml-4.3.3 netCDF4-1.5.1.2 numpy-1.16.4 pendulum-2.0.4 pygeoif-0.7 pyproj-2.2.0 pytz-2019.1 pytzdata-2019.1 regex-2019.6.2 typing-3.6.6
++ python -c 'import json, sys; print json.load(sys.stdin)['\''jenkins://imos-binary/aodncore_prod?pattern=^.*\.whl$'\'']['\''local_file'\'']'
++ aodnfetcher -c .python-aodndata-download-cache 'jenkins://imos-binary/aodncore_prod?pattern=^.*\.whl$'
__init__() got an unexpected keyword argument 'id'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.7/json/__init__.py", line 291, in load
    **kw)
  File "/usr/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
+ pip install
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
ERROR: You must give at least one requirement to install (see "pip help install")

```

**Comment by lwgordonimos** [comment]
Delete the ".python-aodndata-download-cache" and let it rebuild the cache.

The caching code of the fetcher was changed recently, and the cacheindex.json file is incompatible when the new code tries to use an old cache.

**Comment by mhidas** [comment]
:+1:   That works, thanks!

**Comment by lwgordonimos** [comment]
Ideally it would have pruned these out automatically, but it's such an edge case going forward that it's not really worth fixing in code now!

**Pull request merged**